django/db/models/query.py

```diff
@@ -1120,36 +1118,27 @@ def _merge_known_related_objects(self, other):
         for field, objects in other._known_related_objects.items():
             self._known_related_objects.setdefault(field, {}).update(objects)
 
-    def _prepare(self, field):
-        if self._fields is not None:
-            # values() queryset can only be used as nested queries
-            # if they are set up to select only a single field.
-            if len(self._fields or self.model._meta.concrete_fields) > 1:
-                raise TypeError('Cannot use multi-field values as a filter value.')
-        elif self.model != field.model:
-            # If the query is used as a subquery for a ForeignKey with non-pk
-            # target field, make sure to select the target field in the subquery.
-            foreign_fields = getattr(field, 'foreign_related_fields', ())
-            if len(foreign_fields) == 1 and not foreign_fields[0].primary_key:
-                return self.values(foreign_fields[0].name)
-        return self
-
-    def _as_sql(self, connection):
-        """
-        Returns the internal query's SQL and parameters (as a tuple).
-        """
-        if self._fields is not None:
+    def _prepare_as_filter_value(self):
```

Comment:
I'm usually fairly conscious of higher level abstractions accessing very detailed properties or methods from lower down the stack. I would prefer a method within the `sql/query.py` Query object so that other implementations that may or may not yet exist have access to override this behaviour.

I'd consider pushing everything from retrieving the inner query into a method and returning on that.

```
return obj.query.as_subquery_filter(obj)
```

Method name and args probably need work but that's the kind of thing I'm considering.